certora/specs/ERC20Invariants.spec

```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+methods {
+    function totalSupply() external returns uint256 envfree;
+    function balanceOf(address) external returns uint256 envfree;
+    function delegatee(address) external returns address envfree;
+    function delegatedVotingPower(address) external returns uint256 envfree;
+    function upgradeToAndCall(address, bytes) external => NONDET DELETE;
+}
+
+// Paramater for any address that is not the zero address.
+persistent ghost address A {
+    axiom A != 0;
+}
+
+// Partial sum of delegated votes to parameterized address A.
+// sumOfvotes[x] = \sum_{i=0}^{x-1} balances[i] when delegatee[i] == A;
+ghost mapping(mathint => mathint) sumsOfVotes {
+    init_state axiom forall mathint account. sumsOfVotes[account] == 0;
+}
+
+// Ghost copy of DelegationTokenStorage._delegatee for quantification.
+ghost mapping(address => address) ghostDelegatee {
+    init_state axiom forall address account. ghostDelegatee[account] == 0;
+}
+
+// Slot is DelegationTokenStorage._delegatee.
+hook Sload address delegatee (slot 0x669be2f4ee1b0b5f3858e4135f31064efe8fa923b09bf21bf538f64f2c3e1100)[KEY address account] {
+    require ghostDelegatee[account] == delegatee;
+}
+
+// Slot is DelegationTokenStorage._delegatee.
+hook Sstore (slot 0x669be2f4ee1b0b5f3858e4135f31064efe8fa923b09bf21bf538f64f2c3e1100)[KEY address account] address delegatee (address delegateeOld) {
+    mathint changeOfAccountVotesForA;
+    // Track delegation changes from the parameterized address.
+    if (delegateeOld == A && delegatee != A) {
+        require changeOfAccountVotesForA == - ghostBalances[account];
+    // Track delegation changes to the pramaeterized address.
+    } else if (delegateeOld != A && delegatee == A) {
+        require changeOfAccountVotesForA == ghostBalances[account];
+    } else {
+        require changeOfAccountVotesForA == 0;
+    }
+    // Update partial sums for x > to_mathint(account)
+    havoc sumsOfVotes assuming
+        forall mathint x. sumsOfVotes@new[x] ==
+        sumsOfVotes@old[x] + (to_mathint(account) < x ? changeOfAccountVotesForA : 0);
+    // Update ghost copy of DelegationTokenStorage._delegatee.
+    ghostDelegatee[account] = delegatee;
+}
+
+// Ghost variable to hold the sum of balances.
```

Comment:
let's explain better

Reply:
Fixed in https://github.com/morpho-org/morpho-token/pull/102/commits/ecf40702f9934113f353ee0e5c0e59bba9fa048e.